python/lsst/meas/astrom/matchPessimisticB.py

```diff
@@ -432,73 +433,94 @@ def _doMatch(self, refCat, sourceCat, wcs, refFluxField, numUsableSources,
+                    # can't find an optimistic match on our first try with the
+                    # tolerances as found in the previous match-fit,
+                    # the match we found in the last iteration was likely bad. We
+                    # append the bad match's index to the a list of
+                    # patterns/matches to skip on subsequent iterations.
+                    match_tolerance.failedPatternList.append(
+                        match_tolerance.lastMatchedPattern)
+                    match_tolerance.lastMatchedPattern = None
+                    maxShiftArcseconds = \
+                        self.config.maxOffsetPix * wcs.pixelScale().asArcseconds()
+                    continue
+                elif len(matcher_struct.match_ids) > 0:
+                    # Match found, save a bit a state regarding this pattern
+                    # in the match tolerance class object and exit.
+                    match_tolerance.maxShift = \
+                        matcher_struct.shift * afwgeom.arcseconds
+                    match_tolerance.lastMatchedPattern = \
+                        matcher_struct.pattern_idx
+                    match_found = True
+                    break
+            if match_found:
                 break
 
+        # If we didn't find a match, exit early.
+        if not match_found:
+            return pipeBase.Struct(
+                matches=[],
+                match_tolerance=match_tolerance,
+            )
+
         # The matcher returns all the nearest neighbors that agree between
         # the reference and source catalog. For the current astrometric solver
         # we need to remove as many false positives as possible before sending
         # the matches off to the solver.
-        distances_arcsec = np.degrees(matcher_struct.distances_rad) * 3600
+        distances_arcsec = np.degrees(matcher_struct.distances_rad) * 3600 
+        clip_max_dist = np.max(
+            (sigmaclip(distances_arcsec, low=100, high=2)[-1],
+             self.config.minMatchDistPixels * wcs.pixelScale().asArcseconds())
+        )
         # We pick the largest of the the unsoftened maxMatchDistArcSec or
         # a user specified distance in pixels. This prevents the
         # AstrometryTask._matchAndFitWCS from over-fitting to a small number of
         # objects and also allows the WCS fitter to bring in more matches as
         # the WCS fit improves.
-        dist_cut_arcsec = np.max(
-            (self.config.minMatchDistPixels * wcs.pixelScale().asArcseconds(),
-             maxMatchDistArcSec)
-        )
+        if not np.isfinite(clip_max_dist):
+            clip_max_dist = maxMatchDistArcSec
+        
+        if clip_max_dist < maxMatchDistArcSec and \
+           len(distances_arcsec[distances_arcsec < clip_max_dist]) < \
+           minMatchedPairs:
+            dist_cut_arcsec = maxMatchDistArcSec
+        else:
+            dist_cut_arcsec = np.min((clip_max_dist, maxMatchDistArcSec))
```

Comment:
These changes make the comment above untrue — you are not picking the largest of anything, but the minimum of `clip_max_dist` and `maxMatchDistArcSec` (except when you aren't).

Reply:
The comment has been edited.